index.js

```diff
@@ -155,179 +145,142 @@ function readProtocolData (reader) {
-
+function readPrepare (reader) {
   const transferId = uuidParse.unparse(reader.read(16))
-  const amount =twoNumbersToString(reader.readUInt64())
+  const amount = twoNumbersToString(reader.readUInt64())
   const executionCondition = base64url(reader.read(32))
   const expiresAt = readGeneralizedTime(reader)
   const protocolData = readProtocolData(reader)
-
-  return { requestId, transferId, amount, executionCondition, expiresAt, protocolData }
-}
-
-function serializeFulfill ({ transferId, fulfillment }, requestId, protocolData) {
-  const transferIdBuffer = Buffer.from(transferId.replace(/-/g, ''), 'hex')
-  const fulfillmentBuffer = Buffer.from(fulfillment, 'base64')
-  const writer = new Writer()
-
-  writer.write(transferIdBuffer)
-  writer.write(fulfillmentBuffer)
-  writeProtocolData(writer, protocolData)
-
-  return writeEnvelope(TYPE_FULFILL, requestId, writer.getBuffer())
+  return { transferId, amount, executionCondition, expiresAt, protocolData }
 }
 
-function deserializeFulfill (buffer) {
-  const { requestId, data } = readEnvelope(buffer)
-  const reader = new Reader(data)
-
+function readFulfill (reader) {
   const transferId = uuidParse.unparse(reader.read(16))
   const fulfillment = base64url(reader.read(32))
   const protocolData = readProtocolData(reader)
-
-  return { requestId, transferId, fulfillment, protocolData }
-}
-
-function serializeReject ({ transferId, rejectionReason }, requestId, protocolData) {
-  const transferIdBuffer = Buffer.from(transferId.replace(/-/g, ''), 'hex')
-  const rejectionReasonBuffer = maybeSerializeIlpError(rejectionReason)
-
-  const writer = new Writer()
-  writer.write(transferIdBuffer)
-  writer.write(rejectionReasonBuffer)
-  writeProtocolData(writer, protocolData)
-
-  return writeEnvelope(TYPE_REJECT, requestId, writer.getBuffer())
+  return { transferId, fulfillment, protocolData }
 }
 
-function deserializeReject (buffer) {
-  const { requestId, data } = readEnvelope(buffer)
-  const reader = new Reader(data)
-
+function readReject (reader) {
   const transferId = uuidParse.unparse(reader.read(16))
   const rejectionReason = readIlpError(reader)
   const protocolData = readProtocolData(reader)
-
-  return { requestId, transferId, rejectionReason, protocolData }
+  return { transferId, rejectionReason, protocolData }
 }
 
-function serializeMessage (requestId, protocolData) {
-  const writer = new Writer()
+function deserialize (buffer) {
+  const envelopeReader = Reader.from(buffer)
 
-  writeProtocolData(writer, protocolData)
+  const type = envelopeReader.readUInt8()
+  const requestId = envelopeReader.readUInt32()
+  const dataBuff = envelopeReader.readVarOctetString()
+  const reader = new Reader(dataBuff)
 
-  return writeEnvelope(TYPE_MESSAGE, requestId, writer.getBuffer())
-}
+  let data
+  switch (type) {
+    case TYPE_ACK:
+    case TYPE_RESPONSE:
+    case TYPE_MESSAGE:
+      data = readProtocolData(reader) // see https://github.com/interledger/rfcs/issues/284
```

Comment:
That is a breaking change. Can we make this for now non-breaking by changing this line to `data = {protocolData: readProtocolData(reader)`.

Also see this related PR to make the specs match the implementation: https://github.com/interledger/rfcs/pull/285

Reply:
A bug fix is by definition breaking the behavior for people who relied on that bug being present :)

Reply:
This wasn't a bug in the first place, but intentional to avoid the problems outlined in https://github.com/interledger/rfcs/issues/284.

Lets merge https://github.com/interledger/rfcs/pull/285 an be done with it.